shared_stack.bash

```diff
@@ -0,0 +1,214 @@
+       # shellcheck disable=SC2046
+       conda list --json | grep mypy > /dev/null \
+         || mamba install -y -c conda-forge --no-update-deps $(cat "$STATE/developer.txt")
+     fi
+     $dryrun eups distrib install -t "$tag" "$PRODUCT"
+
+     # Rewrite load scripts and write useful files
+     $dryrun cp loadLSST.sh tmp.loadLSST.sh
+     if [ -n "$dryrun" ]; then
+       echo "echo . $ROOT/loadLSST.sh ... > loadLSST.sh"
+       echo "echo conda activate ... > switch.sh"
+       echo "echo ... > env_name"
+     else
+       echo "LSST_CONDA_ENV_NAME=$LSST_CONDA_ENV_NAME . \"$ROOT/loadLSST.sh\"" > loadLSST.sh
+       echo "conda activate $LSST_CONDA_ENV_NAME" > switch.sh
+       echo "$LSST_CONDA_ENV_NAME" > env_name
+     fi
+  ) && [ ! -d "$dir" ] && $dryrun mv "${dir}.$tmp" "$dir"
+  # Clean up if we failed or lost the race
+  [ -d "${dir}.$tmp" ] && rm -rf "${dir}.$tmp"
+done
+rm $tmp/tags
+
+# List the installed tags
+(
+  cd "$ROOT/tag"
+  ls -dr d_* > "$STATE/$tmp/d_list" || true
+  ls -dr w_* > "$STATE/$tmp/w_list" || true
+)
+
+# Make the latest links if needed
+for type in d w; do
+  [ -f "$tmp/${type}_list" ] || continue
+  latest=$(head -n 1 "$tmp/${type}_list")
+  link="$ROOT/${type}_latest"
+  if [[ ! -L "$link" || "$(readlink "$link")" != "$ROOT/tag/$latest" ]]; then
+    $dryrun rm -f "$link" && $dryrun ln -s "$ROOT/tag/$latest" "$link"
+  fi
+done
+# Update loadLSST.sh if needed
+tmp_load="$ROOT/w_latest/tmp.loadLSST.sh"
+if [ -n "$dryrun" ]; then
+  echo "[ -f \"$tmp_load\" ] && mv \"$tmp_load\" \"$ROOT/loadLSST.sh\""
+elif [ -f "$tmp_load" ] && ! diff -q "$ROOT/loadLSST.sh" "$tmp_load" > /dev/null 2>&1; then
+  [ -f "$ROOT/loadLSST.sh" ] && mv "$ROOT/loadLSST.sh" "$ROOT/loadLSST.sh.bak"
+  mv "$tmp_load" "$ROOT/loadLSST.sh"
+fi
+$dryrun rm -f "$ROOT"/tag/*/tmp.loadLSST.sh
+# Update current tag to latest weekly
+# Doesn't handle all cases (an old weekly is installed as the latest in an
+# even older environment), but should be good enough
+if [ -f "$tmp/w_list" ]; then
+  (
+    latest=$(head -n 1 "$tmp/w_list")
+    set -x
+    $dryrun source "$ROOT/tag/$latest/loadLSST.sh"
+    $dryrun eups tags --clone="$latest" current
+  )
+fi
+
+#########
+# Cleanup
+#########
+
+for type in d w; do
+  [ -f "$tmp/${type}_list" ] || continue
+  keep=KEEP_$type
+  for tag_del in $(tail -n "+${!keep}" "$tmp/${type}_list"); do
+      # Add to deleted list so this tag doesn't come back
+      if [ -n "$dryrun" ]; then
+        echo "echo $tag_del >> tags.deleted"
+      else
+        echo "$tag_del" >> tags.deleted || continue
+      fi
+      # Rename the directory so others can't use it
+      del_dir="$ROOT/tag/del.$tag_del"
+      $dryrun mv "$ROOT/tag/$tag_del" "$del_dir" || continue
+      (
+        set +x
+        [ -n "$dryrun" ] && echo "source $del_dir/loadLSST.sh"
+	# shellcheck disable=SC1091
+        source "$del_dir/loadLSST.sh"
+        set -x
+        eups list -t "$tag_del" | while read -r pkg ver tag1 tagn; do
```

Comment:
Can more than one tag be in a directory?

Reply:
The tag directories don't contain stacks.  They only contain pointers to the conda environment (three different kinds).  With lsstinstall, the conda environment itself contains the stack.  And it can contain multiple stack tags that are all consistent with the environment.  So that's why `-t $tag` needs to be specified here.

Reply:
I'm sorry this was meant to apply more to what is below. I think I completely have everything mapped together, can eups stacks contain more than one tag in them?